### ISSUE TITLE
Header issue with borderless Bootstrap table

### ISSUE THREAD

**Issue by mweirauch** [DataTables/FixedColumns#57]
When using a borderless bootstrap table, only the left fixed column has a bottom margin. The non-fixed right columns don't. I dunno yet, what change caused this, as this used to work a while ago IIRC.

The `table-bordered` class is suffixed with an `x` for easy toggling.

```html
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01//EN"
"http://www.w3.org/TR/html4/strict.dtd">
<html>
<head>
  <meta http-equiv="content-type" content="text/html; charset=utf-8">
  <title>DataTables Bootstrap 3 example</title>
  <link rel="stylesheet" type="text/css" href=
  "https://cdn.datatables.net/t/bs-3.3.6/jqc-1.12.0,dt-1.10.11,fc-3.2.1/datatables.min.css">
  <script type="text/javascript" src=
  "https://cdn.datatables.net/t/bs-3.3.6/jqc-1.12.0,dt-1.10.11,fc-3.2.1/datatables.min.js">
  </script>
  <script type="text/javascript">
        /*<![CDATA[*/
        $(document).ready(function() {
    var table = $('#example').DataTable( {
        scrollX:        true,
        scrollCollapse: true,
        paging:         false,
        fixedColumns:   true
    } );
  } );
        /*]]>*/
  </script>
</head>
<body>
  <div class="container">
    <table id="example" class="table table-striped table-borderedx table-condensed nowrap" cellspacing="0" width=
    "100%" summary="example">
      <thead>
        <tr>
          <th>First name</th>
          <th>Last name</th>
          <th>Position</th>
          <th>Office</th>
          <th>Age</th>
          <th>Start date</th>
          <th>Salary</th>
          <th>Extn.</th>
          <th>E-mail</th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>Tiger</td>
          <td>Nixon</td>
          <td>System Architect</td>
          <td>Edinburgh</td>
          <td>61</td>
          <td>2011/04/25</td>
          <td>$320,800</td>
          <td>5421</td>
          <td>t.nixon@datatables.net</td>
        </tr>
        <tr>
          <td>Garrett</td>
          <td>Winters</td>
          <td>Accountant</td>
          <td>Tokyo</td>
          <td>63</td>
          <td>2011/07/25</td>
          <td>$170,750</td>
          <td>8422</td>
          <td>g.winters@datatables.net</td>
        </tr>
        <tr>
          <td>Ashton</td>
          <td>Cox</td>
          <td>Junior Technical Author</td>
          <td>San Francisco</td>
          <td>66</td>
          <td>2009/01/12</td>
          <td>$86,000</td>
          <td>1562</td>
          <td>a.cox@datatables.net</td>
        </tr>
      </tbody>
    </table>
  </div>
</body>
</html>
```

**Comment by DataTables** [comment]
Thanks for the code. I've put it into [a running test case](http://live.datatables.net/nunusifi/1/edit).

I'm sure I'm being daft, but could you clarify the issue when the table is borderless? It appears to render as expected for me - or am I missing an artifact?

**Comment by mweirauch** [comment]
Sorry for the late reply, Gmail sorted it away :/

Attached a screenshot of Firefox, Epiphany and Chrome on Linux with the initial test. But livetest shows the same issues. Issues marked in red. Epiphany using a newer WebKit-Version apparently does not expose the issue.

I think the issues with Chrome came with the recent 50.0 update.

Safari 9.1 on Mac is ok.

![datatables-header-bug](https://cloud.githubusercontent.com/assets/1134031/15206257/0b9cdb0e-181e-11e6-9837-b17d6c579298.png)

**Comment by DataTables** [comment]
Do any of the examples on [the DataTables site](https://datatables.net/extensions/fixedcolumns/examples/) show the issue for you? I've just taken a look with Chrome 50 and it appears to work okay.

**Comment by mweirauch** [comment]
Then it would only be the [bootstrap styling example](https://datatables.net/extensions/fixedcolumns/examples/styling/bootstrap.html).

If I manipulate/drop all `table-bordered` classes from the 4 (?) tables generated with datatables all looks fine though. The issue seems to only show up if the initial (non-datatables-modified) table is borderless by default.

Chrome version on Linux is Version 50.0.2661.94 (64-bit). At home (Windows 10) the Chrome major is 50.x as well. Not sure if the minors are exactly the same. But it shows the same erroneous header bottom border behavior.

**Comment by mweirauch** [comment]
Just played with the live example: Whenever I disable the "scrollX" prerequisite (rendering FixedColumns non-functioning IIRC) all header cell bottom borders are drawn, not just the one of the fixed column. Using two or more fixed columns shows their bottom border. All "floating" columns miss the bottom border.

As reference, a default non-bordered Bootstrap table looks like [this](http://getbootstrap.com/css/#tables-striped).

**Comment by mweirauch** [comment]
Just had a look on a Mac with:
* Safara 9.1: all good
  * User-Agent: `Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_5) AppleWebKit/601.5.17 (KHTML, like Gecko) Version/9.1 Safari/601.5.17`
* Firefox 46.0.1: same top right "square" artifact as on Linux
* Chrome 50.0.2661.94: "floating" table headers are missing bottom-border as on Linux

The Safari User-Agent exposes a newer WebKit-Version (601) as does Epiphany (602). Whereas Chrome uses 537.

To summarize: Same Firefox and Chrome issue on 3 platforms. If this is a WebKit 537 regression and fixed in later version as seen in Epiphany or Safari is beyond my knowledge.

**Comment by DataTables** [comment]
Thanks for the details. I'll try to take a look later today. Worth noting that Chrome doesn't use Webkit any more - the forked it to create Blink a while back, so Webkit and Blink are slowly diverging. It is quite possible there will be bugs in one and not the other.

**Comment by DataTables** [comment]
Interestingly this isn't caused by code in FixedColumns, but rather by that in DataTables core. Bootstrap uses `border-collapse: collapse`, but DataTables requires `border-collapse: separate` since with `collapse` is is simply impossible to perform the column width calculations required for scrolling tables. That difference was causing what you could see - the non-fixed columns were borderless because they were `collapse`, but due to the CSS priority, the fixed columns wasn't...

I've [committed the fix](https://github.com/DataTables/DataTablesSrc/commit/01d73effb1bb4c25ef5a0bf51689e30316832b7b) and FixedColumns works as it should now.

**Comment by mweirauch** [comment]
Thank you very much! Chrome issue is fixed. Looks as it should.

Unfortunately the Firefox 46.0.1 issue still remains. What on the screenshot above looks like drawing artifact is in fact a scrollbar on a Windows host. It's also visible in the live demo.

Shall I open another ticket? If so, for which project? Core or FixedColums?

**Comment by DataTables** [comment]
I've put this little example together for FixedColumns with Bootstrap borderless tables: http://live.datatables.net/yalasame/1/edit . It appears to work okay in Firefox. Does it work for you?

**Comment by mweirauch** [comment]
From a naive point of view, it looks like a scrollbar (system styled) at the rightmost position in the header row in Firefox. Linux this time. Looks the same on Windows, though.

![datatables-bootstrap-borderless-header-bug-firefox](https://cloud.githubusercontent.com/assets/1134031/15500311/7a329f34-21a8-11e6-9af1-06629f315aef.png)

As if FF determines the box needs scrolling and displays the scrollbar (in FF without arrows here on Gnome) but there's basically nothing to scroll or so little space to scroll at all, that the scroll nob isn't drawn at all.

But I might totally mistake that grey square as a scrollbar, when it's something totally different in reality.
